### ISSUE TITLE
Reduce build size by excluding dev packages

### ISSUE THREAD

**Issue by SimonSimCity** [meteor/meteor#10902]
When running `meteor build --server-only` on a clean project (just freshly created by the command `meteor create test-project` I end up with a folder of a size of `118mb` - of which `74mb` is taken up by `programs/server/npm/node_modules/meteor/babel-compiler`. I'm quite unsure if the rest wouldn't be needed, but if we could get out the `babel-compiler` here, that would decrease the size of our final bundle considerably.

**Comment by faburem** [comment]
Is there any update on this? I think this could be a huge benefit especially for docker images. If there is anything we can do to push this forward - I am happy to help.

**Comment by SimonSimCity** [comment]
Just as a small sidenote: Also the packages `ecmascript-runtime-server` and `ecmascript-runtime-client` both have the package `core-js` two times in their current release. Once as `core-js` and once as `.core-js-garbage-*.*`. Maybe it's a good solution to somehow prevent the inclusion of such packages in a release on atmosphere.

**Comment by afrokick** [comment]
Is it safe to remove the Babel-compiler folder? I’ll try to create PR to solve this issue

**Comment by afrokick** [comment]
Not stale

**Comment by crapthings** [comment]
same here

there's a typescript module but i don't even use it

![image](https://user-images.githubusercontent.com/1147704/113399097-3b2aad80-93d2-11eb-827f-45af5df97138.png)

**Comment by distalx** [comment]
Is it necessary to include these node_modules in their entirety? can we get by, by adding only `dist` code? some modules also have their documentation and examples included with them. which are of no use when we use them in production.

**Comment by TimHeckel** [comment]
Yeah - not stale.

**Comment by ensemblebd** [comment]
I believe the logic the engineers meant this for, was to ensure that precisely what is used locally is what ends up on the server.
Since npm can get pretty wonky back in the day before package-lock.json.
Even still, package-lock.json cross platform is a serious issue too. God save the man who must use node-gyp cross platform.

But i decided to disable it anyway. I don't think meteor would be proud of this, I'm not even proud of it, but i have shaved more than 50% of build time and data xfer amount by doing this. 

**And though it works for me, you should verify it for yourself for safety, especially in a production environment.** 
And know that you MUST run `npm install` on the server. 
**Warning**: I have not tested this change on bundler running for android or ios packages. Only SERVER. It works for server bundle. I can't vouch the other two right now, too tired after 50+ hour shifts.

_(replace all relevant dirs, for your version, your operating system, etc)_
Windows: `C:\Users\<you>\AppData\Local\.meteor\packages\meteor-tool\2.3.2\mt-os.windows.x86_64\tools\isobuild\bundler.js`
Mac: `~/.meteor/packages/meteor-tool/2.3.2/mt-os.osx.x86_64/tools/isobuild/bundler.js`
Line # 2317 (create new line, first line of inline function):
`return;`
And so it shall read:
```
      var nodeModulesDirectories = Object.create(null);

      _.each(self.nodeModulesDirectories || [], function (nmd) {
        return;
        // We need to find the actual file system location for the node modules
        // this JS Image uses, so that we can add it to nodeModulesDirectories
        var modulesPhysicalLocation;
        
        ....
```

they could easily make this a parameter for the tool, or maybe a hidden env-var for the advanced folks like us who are at our wits-end.
Again, haven't tested mobile. So in theory it's possible it may need some logistics other than just a blind return statement.

**Comment by StorytellerCZ** [comment]
@ensemblebd thanks for the workaround. Sadly as you mentioned this does not look like something that we can integrate into Meteor itself as it is solution specific.

**Comment by ensemblebd** [comment]
@StorytellerCZ  understood. Do you think it may be permissible, if perhaps a changeset was produced whereby:
- a new build parameter was offered:   --serverPackages [embed | manual]    
- default to "embed" which is natural (current)) behaviour
- "manual" mode would: 
  - skip packaging npm local packages into server bundle
  - instead would include package-lock.json into the bundle
  - warn user to run npm-install on the server during build output

I'd be happy to try and find some time and produce a pull request on such a thing.
But I don't want to waste time if you think it would not be permissible.  Let me know your thoughts thank you!

**Comment by StorytellerCZ** [comment]
That is potential one solution, though that one seem too much extra work and would require addition of an action on deployment which I don't think is optimal.
I think the first step should be fixing the packages so that server only packages don't create anything for client bundles.
@filipenevola and @renanccastro might offer additional input.

**Comment by afrokick** [comment]
@StorytellerCZ is it good idea to convert this issue into discussion?

**Comment by afrokick** [comment]
@renanccastro Done https://github.com/meteor/meteor/discussions/11912

**Comment by nachocodoner** [comment]
I encountered this issue after seeing that `babel-compiler` became even heavier following the [SWC integration in Meteor 3.3](https://forums.meteor.com/t/faster-builds-in-meteor-3-3-modern-build-stack-with-swc-and-bundler-optimizations/63697).

I’m glad this has been discussed in the past. I kept doubts that babel-compiler runs at runtime, since it shouldn’t. Your confirmation in this issue gives me more confidence to continue working on a solution.

I want to try removing from the final bundle any packages marked as `devLike`. We already support `testOnly`, `prodOnly`, and `debugOnly` for packages, so I just need to add `devLike`, which would be omitted in production.

[I’ve enabled the changes to `meteor profile`](https://github.com/meteor/performance/pull/18) to quickly check final bundle sizes, so we can verify the savings from marking `devLike` packages.

I’ll share the PR once it’s ready.

**Comment by zodern** [comment]
> I want to try removing from the final bundle any packages marked as devLike

It might not have the best name, but `debugOnly` does this (it is included in development builds and excluded from production builds).

The challenge with `babel-compiler` is it is used by production packages in most Meteor apps. I partially fixed it in Meteor 3 by removing it as a production dependency of the `ecmascript` package, but it is still a dependency of the `shell-server` package that most apps use.

**Comment by nachocodoner** [comment]
> It might not have the best name, but debugOnly does this (it is included in development builds and excluded from production builds).

Yes, `debugOnly` seems to do that! It caused confusion for me. devOnly or just dev would be more descriptive. But great to see it does that exactly.

> but it is still a dependency of the shell-server package that most apps use.

Do you think `shell-server` is the only remaining dependent package, or are there others? I'll double check myself anyway.

I will think to make it independent. Have you any ideas? [It's used to eval code on-fly](https://github.com/meteor/meteor/blob/devel/packages/shell-server/shell-server.js#L238-L249). Maybe I can adapt `shell-server` to use SWC for code evaluation as well, or a thinner alternative, which should be slightly smaller than the current babel-compiler.

Also, I don’t use this shell-server in production and I’m not sure how often it’s used for others. Devs could exclude it from their apps when no used and even not add this package to new Meteor apps automatically.

---

**Edit:**

Using `debugOnly` prevents `meteor build` from working, as build plugins relying on those dependencies aren’t applied. An option that guarantees these packages are excluded from the final bundle, either by not including them or by stripping them out after build, would solve this. I'll prepare a PR with a proposal.

**Comment by nachocodoner** [comment]
What’s the status of shell-server? [The official docs](https://docs.meteor.com/cli/#meteorshell) and [some forum posts](https://forums.meteor.com/t/access-a-production-server-with-meteor-shell/30548/2?u=nachocodoner) say it’s a dev-only package (`meteor run`). I’ve only seen it enabled for production or deployed builds via the [prod-shell package](https://github.com/qualialabs/prod-shell), and I’m not sure it’s compatible with Meteor 3.

If shell-server is truly a dev dependency, we can skip it in the production bundle along with babel-compiler. We could migrateprod-shell into core to enable it in production while keeping it optional. Maybe have it depend on the [@meteorjs/babel](https://www.npmjs.com/package/@meteorjs/babel) npm package instead of babel-compiler, so we can drop babel-compiler for production bundle as mark it as devOnly.

**Comment by nachocodoner** [comment]
This was already delivered in Meteor 3.4.

You can mark packages as [devOnly ](https://docs.meteor.com/packages/3.writing-atmosphere-packages.html#dev-only-packages) or add specific [Npm.devDepends](https://docs.meteor.com/api/package.html#PackageNpm-devDepends) so the whole package or selected deps are not included in the final production bundle. This is already applied to some core bundles.

Nothing else is needed to benefit from this change, just upgrade to Meteor 3.4.
